// sass-lint-auto-fix/dist/resolvers/space-before-bang.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const gonzales_pe_sl_1 = require("gonzales-pe-sl");
const base_resolver_1 = __importDefault(require("./base-resolver"));
class SpaceBeforeBang extends base_resolver_1.default {
    fix() {
        const { ast } = this;
        ast.traverseByTypes(['important', 'default', 'global', 'optional'], (_, index, parent) => {
            const prev = parent.content[index - 1] || {};
            const isSpace = prev.type === 'space';
            if (this.shouldAddSpaceBeforeBang()) {
                if (!isSpace) {
                    const spaceNode = gonzales_pe_sl_1.createNode({
                        type: 'space',
                        syntax: ast.syntax,
                        content: ' ',
                    });
                    parent.content.splice(index, 0, spaceNode);
                }
            }
            else if (isSpace) {
                parent.removeChild(index - 1);
            }
        });
        return ast;
    }
    shouldAddSpaceBeforeBang() {
        return this.parser.options.include;
    }
}
exports.default = SpaceBeforeBang;
